README.md
# Appkit

Packages to do common stuff.

# Log

Provide a default logfmt-style logger (using go-kit/log/levels) for your app.

```go
logger := log.Default()

err := someAction()
logger.Error().Log(
    "msg", "Something happened",
    "during", "someAction",
    "err", err
) // => level=error msg="Something happened", during=someAction err=<error as string>
```

You can pre-set common fields:

```go
l := logger.With("context", "README")

logger.Info("msg", "the message") // => level=info context=README msg="the message"
```


Supported levels:

* `Debug`
* `Info`
* `Warn`
* `Error`
* `Crit`

## Field recommendations

* `msg`: human-readable version of the log, eg. `"fmt.Sprintf("error loading config: %v", err)`, `"registering gorm callbacks"`.
* `context`: what the system was doing when the info was logged, eg `oauthprovider.endpoint`, `dbmigration`.
* `err`: the error being logged
* `during`: the function that returned an error or induced the log, eg `influxdb.Client.Write`

Use others that are relevant to your app or domain.

## No operation logger for testing

Provide a logger that doesn't do anything. This is quite useful for testing.

```go
logger := log.NewNopLogger()
```

## Set flag `APPKIT_LOG_HUMAN` for better dev experience

`export APPKIT_LOG_HUMAN=true` Will make the logger outputs to a format that is easily to read for developers.


# Server

## HTTP Listener

Helper for starting a HTTP server configured with a `log.Logger`. Provides `Config` and `ListenAndServe`.

## Middleware

* `ETag`: will md5 your response body and include the hash in the `ETag` HTTP header. If client provided the same ETag in `If-None-Match` HTTP header, will return `304 Not Modified` and discard the response.

  Does nothing (ie, passthrough) with non-`GET` requests and non-`200 OK` responses.

* `LogRequest`: logs incoming HTTP requests with `log`. Will log start and end of request. Uses logger from request `context.Context`, so any fields set with `log.Logger.With` will be included in the log.

* `Recovery`: recovers `panic` in HTTP handlers, sends `500 Internal Server Error` to the client, and re-`panic`s the recovered error.

* `DefaultMiddleware`: Default middleware stack: request -> record HTTP status -> trace -> log -> recover.

* `Compose`: helper to chain middleware together.

# DB

Helper for opening a `gorm.DB` connection configured with a `log.Logger`. Provides `Config` and `New`.

# [Monitoring](monitoring/README.md)

A basic interface for monitoring request times and other arbitrary data, and recording data into InfluxDB.

# [Error Notification](errornotifier/README.md)

Interface for pushing panics and arbitrary errors into error logging systems. Provides implementation for Airbrake monitoring.

# [Sessions](sessions/README.md)

This package is a wrapper of [gorilla/sessions](https://www.github.com/gorilla/sessions) to fix the potential [memory leaking problem](https://qortex.com/theplant#groups/560b63da8d93e34b8500da28/entry/58a297e98d93e316d10328f3).

# Contexts

Context wrappers and http.Handler middleware to setup and use various `context.Context`s.

* `RequestTrace`: generate unique id for each HTTP request. Useful for tracing everything that happens due to a single request. Used by `Logger`

* `Logger`: makes a given logger available via `context.Context`. Integrated with `RequestTrace` to add the request trace ID to anything logged via the context, if request is being traced. Used by `server.LogRequest` middleware.

* `HTTPStatus`: records the HTTP status code for the request. Used by `LogRequest` to record the final response HTTP status code.

* `Gorm`: make a `gorm.DB` available via `context.Context`.


## Naming Style

For an "ABC" context:

* `ABC` will extract the "value" from a `context.Context`. Generally returns the value and a `bool` to indicate whether the context actually had any value.
* `WithABC` is `http.Handler` middleware that will enable "ABC" in a HTTP handler.
* `ABCContext` will wrap a `context.Context` and provide a new context that can be passed to `ABC`.
* `MustGetABC` is a wrapper around `ABC` that will `panic` when `ABC` would return false. Useful when you *need* the context value and the only way you'd handle a missing value would be to `panic`.

# [Encrypted Box](encryptedbox/README.md)

Secret Box provides a simple interface for encryption of data for storage at rest.

It is implemented as a simple wrapper around `golang.org/x/crypto/nacl/secretbox` that takes care of handling the nonce.

# [Tracing](tracing/README.md)

Tracing supports distributed tracing of requests by linking together
all of the parts of work that go into fulfilling a request.

For example, with a HTML front-end talking to back-end HTTPS APIs, it
will link the original front-end request with any/all HTTP requests
made to the back-end. Also, it can link together deeper requests made
by the *back-end* to other APIs and services.

For now, It's implemented with [OpenTracing](https://opentracing.io)
and expects to talk to a [Jaeger](https://www.jaegertracing.io)
back-end.

# Credentials

Package to provide single interface for acquiring credentials for apps
running on different platforms. Credentials that can be sourced:

* AWS
* InfluxDB

Places to source credentials from:

* [Vault](https://www.vaultproject.io)
* Local environment

# [Service](service/README.md)

Package to provide a common harness for running HTTP apps, that
configures middleware that we use nearly all the time, and provides a
standard way to configure the different parts of the service.


CHANGELOG.md
# PR [#45](https://github.com/theplant/appkit/pull/45)

* Add log.Start(ctx) to include duration=123.11ms in the following log output that use it.

# PR [#39](https://github.com/theplant/appkit/pull/39)

* Add monitoring InfluxDB client that sources InfluxDB credentials
  from Vault.

* Allow creation of InfluxDB-backed monitoring client with custom
  InfluxDB transport client.

* Introduce public Vault client interface (instead of using
  `github.com/hashicorp/vault/api.Client` directly)

* Make Vault client available from service context

* Allow subscription to Vault (re-)authentication via
  `appkit/credentials/vault.Client.OnAuth`

* Revoke Vault auth lease when service context is closed

# PR [#38](https://github.com/theplant/appkit/pull/38)

* Add support for `VAULT_AUTHN_DISABLED` to stop appkit/service
  automatically attempting to authenticate with Vault.

# PR [#37](https://github.com/theplant/appkit/pull/37)

* Add package appkit/service

# PR [#19](https://github.com/theplant/appkit/pull/19)

* Add Package appkit/credentials

# PR [#36](https://github.com/theplant/appkit/pull/36)

## Breaking Changes

* `errornotifier.NewAirbrakeNotifier` func now returns `io.Closer` to
  wait send data to Aribrake.
* Add a new parameter `context` to `errornotifier.Notifier.Notify` allow
  add context when call `Notify`.

## Changes

* Upgrade Airbrake package.
* Change `airbrakeNotifier.Notify` func to async.
* Add log req_id and span_context to `errornotifier.NotifyOnPanic` and
  `Recover` middleware.

# PR [#33](https://github.com/theplant/appkit/pull/33)

* `monitoring.WithMonitor` middleware now collects `X-Via`
  request header value as `via` tag value.

# PR [#31](https://github.com/theplant/appkit/pull/31)

* Log request id in trace log.
* Log trace id in InfluxDB.
* Add `service-name` config for InfluxDB.

# PR [#30](https://github.com/theplant/appkit/pull/30)

## Breaking Changes

* `monitoring.NewInfluxdbMonitor` now returns a "close" `func()` to
  allow termination of InfluxDB buffer introduced in #25.

## Added

* `server.GoListenAndServe` gives control over shut-down of the HTTP
  server.


## Changed Behaviour

* `server.ListenAndServe` now delegates the HTTP server to this
  function and `server.GoListenAndServe` and *manually* blocks forever
  by deadlocking itself. This preserves the basic behaviour, but it
  might change how `ListenAndServe` responded to OS signals.

* Tracing on monitoring middleware was removed. Tracing now occurs on
  the InfluxDB monitor buffer goroutine that sends batches of points
  to InfluxDB.

# PR [#25](https://github.com/theplant/appkit/pull/25)

* Add batch write (buffer) for monitoring/influxdb.

# PR [#20](https://github.com/theplant/appkit/pull/20)

* Add `tracing` package

* Make `errornotifier` and `monitoring` middleware trace API requests

# PR [#27](https://github.com/theplant/appkit/pull/27)

* Change influxdb client import path. Detail in [issue](https://github.com/influxdata/influxdb/issues/11035)

# PR [#17](https://github.com/theplant/appkit/pull/17)

* Don't log SQL query *values* in Gorm log adapter

# PR [#16](https://github.com/theplant/appkit/pull/16)

* Replace the `sessions.Config` with `sessions.CookieStoreConfig`, the `sessions.CookieStoreConfig` is compatible with `jinzhu/configor`, change the struct to enable `Secure` and `HttpOnly` by default and extend the struct to make it general.

* Add `sessions.NewCookieStore` to easy new a `gorilla/sessions.CookieStore` with `sessions.Config`.

# PR [#14](https://github.com/theplant/appkit/pull/14)

* Add `server.SecureMiddleware` for CSRF/CORS API configuration.

* Add `log.NewTestLogger` for stable log output when testing (ie. no
  timestamps or stack references)

# PR [#12](https://github.com/theplant/appkit/pull/12)

* Upgrade [`influxdb.Client`](https://github.com/influxdata/influxdb/tree/master/client#description) to `v2` version to fix the monitoring/influxdb monitor broken [issue](https://github.com/theplant/appkit/issues/11) due to the old version of `influxdb.Client` being deprecated.

# PR [#9](https://github.com/theplant/appkit/pull/9)

* Fix `contexts.HTTPStatus` to assume that the response is `http.StatusOK`, rather than `0`.

# PR [#6](https://github.com/theplant/appkit/pull/6)

* Add `log.WrapError` to `log`

# PR [#2](https://github.com/theplant/appkit/pull/2)

* Add `encryptedbox` package

# PR [#43](https://github.com/theplant/appkit-private/pull/43)

* Add `APPKIT_LOG_HUMAN` environment flag to output logs with better formatted for human


# PR [#42](https://github.com/theplant/appkit-private/pull/42)

* Add Package appkit/sessions

# [Add support for logging fields via monitoring.Monitor (PR#37)](https://github.com/theplant/appkit-private/pull/37)

## Breaking changes

* [Changed monitoring interface](https://github.com/theplant/appkit-private/pull/37/commits/5dde9fa2bc77527f9760feecdda762864fa0572c#diff-2b8f43b8889cf5d451e1b7e74a89ae74L62) to accept data fields in addition to tags.

## Changed behaviour

* Monitoring middleware now logs `req_id` as a field, rather than a
tag, to avoid generating InfluxDB series with immense
[cardinality](https://docs.influxdata.com/influxdb/v1.2/concepts/glossary/#series-cardinality).

# PR [#35](https://github.com/theplant/appkit-private/pull/35)

## Added

* Add Package appkit/kerrs
* Add `log.WithError` to log to be able to log appkit errors with ease

# PR [#33](https://github.com/theplant/appkit-private/pull/33)

## Added

* [`log.NewNopLogger` function](https://github.com/theplant/appkit/blob/08b478e/log/log.go#L74-L78)
* [`log.Context` function](https://github.com/theplant/appkit/blob/08b478e/log/context.go#L29-L32)

## Fixed

* `log.FromContext` panic when receiving a nil context bug

# PR [#32](https://github.com/theplant/appkit-private/pull/32)

## Added

* [`errornotifier` package](errornotifier/README.md)

# PR [#30](https://github.com/theplant/appkit-private/pull/30)

## Added

* `monitoring` package



log/human.go
package log

import (
	"fmt"
	"os"
	"strings"

	slog "log"

	"time"

	"github.com/go-kit/kit/log"
)

/*
Human is for create a Logger that print human friendly log
*/
func Human() Logger {
	l := log.NewSyncWriter(os.Stdout)
	lg := Logger{
		log.LoggerFunc(func(values ...interface{}) (err error) {
			fmt.Fprint(l, PrettyFormat(values...))
			return
		}),
	}
	var timer log.Valuer = func() interface{} { return time.Now().Format("15:04:05.99") }
	lg = lg.With("ts", timer)

	slog.SetOutput(log.NewStdlibAdapter(lg))

	return lg
}

/*
PrettyFormat accepts log values and returns pretty output string
*/
func PrettyFormat_(values ...interface{}) (r string) {
	var ts, msg, level, stacktrace, sql, sqlValues interface{}
	var shorts []interface{}
	var longs []interface{}
	var isSQL bool

	for i := 1; i < len(values); i += 2 {
		key := values[i-1]
		val := values[i]
		if key == "ts" {
			ts = val
			continue
		}
		if key == "msg" {
			msg = val
			continue
		}

		if key == "level" {
			level = val
			continue
		}

		if key == "stacktrace" {
			stacktrace = val
			continue
		}

		if key == "query" {
			sql = val
			isSQL = true
			continue
		}

		if isSQL && key == "values" {
			sqlValues = val
			continue
		}

		if len(fmt.Sprintf("%+v", val)) > 50 {
			longs = append(longs, fmt.Sprintf("\033[34m%+v\033[39m=%+v", key, val))
			continue
		}

		shorts = append(shorts, fmt.Sprintf("\033[34m%+v\033[39m=%+v", key, val))
	}

	var pvals = []interface{}{}
	if ts != nil {
		pvals = append(pvals, fmt.Sprintf("\033[36m%s\033[0m", ts))
	}

	if msg != nil {
		color := "39"
		level = fmt.Sprintf("%+v", level)
		switch level {
		case "crit":
			color = "35"
		case "error":
			color = "31"
		case "warn":
			color = "33"
		case "info":
			color = "32"
		case "debug":
			color = "90"
		}
		pvals = append(pvals, fmt.Sprintf("\033[%sm%s", color, msg))
	}

	pvals = append(pvals, shorts...)
	if len(longs) > 0 {
		pvals = append(pvals, "\n")
		for _, long := range longs {
			pvals = append(pvals, "          ", long, "\n")
		}
	}

	if sql != nil {
		pvals = append(pvals, fmt.Sprintf("\n            %s", sql), "\n")
		if sqlValues != nil {
			pvals = append(pvals, fmt.Sprintf("           \033[34m%s\033[0m=%s", "values", sqlValues), "\n")
		}
	}

	if stacktrace != nil {
		pvals = append(pvals, fmt.Sprintf("\n%s", stacktrace), "\n")
	}

	return fmt.Sprintln(pvals...)
}

func ex(r map[string][]interface{}, f string) ([]interface{}, bool) {
	v, ok := r[f]
	delete(r, f)
	//	if ok && len(v) > 0 {
	//		return strings.Join(v, "."), true
	//	}
	return v, ok

}

func mapVals(kvs ...interface{}) map[string][]interface{} {
	m := map[string][]interface{}{}
	for i, key := range kvs {
		if i%2 == 1 {
			continue
		}

		v, ok := m[key.(string)]
		if !ok {
			v = []interface{}{}
		}

		v = append(v, kvs[i+1])

		m[key.(string)] = v
	}
	return m
}

var blankTime = strings.Repeat(" ", len(time.StampMilli))

/*
PrettyFormat accepts log values and returns pretty output string
*/
func PrettyFormat(values ...interface{}) string {
	r := mapVals(values...)

	level := "<none>"
	if levelA, ok := ex(r, "level"); ok && len(levelA) > 0 {
		level = levelA[0].(string)
	}

	tsStr := strings.Repeat(" ", len(time.StampMilli))
	if ts, ok := ex(r, "ts"); ok && len(ts) > 0 {
		tsStr = ts[0].(string)
	}

	msg := []interface{}{"<no msg>"}
	if msgI, ok := ex(r, "msg"); ok && len(msgI) > 0 {
		msg = msgI
	}

	colour := "37"
	switch level {
	case "crit":
		colour = "35"
	case "error":
		colour = "31"
	case "warn":
		colour = "33"
	case "info":
		colour = "32"
	case "debug":
		colour = "30"
	}

	output := []string{fmt.Sprintf("\033[%s;1m%s: %s\033[0m (%s)", colour, tsStr, msg[0], level)}

	for _, m := range msg[1:] {
		output = append(output, fmt.Sprintf("\033[%s;1m%s  %s\033[0m", colour, blankTime, m))
	}

	for k, v := range r {
		output = append(output, fmt.Sprintf("  %s=%v", k, v))
	}

	output = append(output, "", "")

	return strings.Join(output, "\n")
}
